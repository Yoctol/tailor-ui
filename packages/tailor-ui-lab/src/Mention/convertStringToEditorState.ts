import {
  ContentState,
  EditorState,
  convertFromRaw,
  convertToRaw,
} from 'draft-js';

const getIndicesOf = (searchStr: string, str: string) => {
  const searchStrLen = searchStr.length;

  if (searchStrLen === 0) {
    return [];
  }

  let startIndex = 0;
  let index = str.indexOf(searchStr, startIndex);
  const indices = [];

  while (index > -1) {
    indices.push(index);
    startIndex = index + searchStrLen;
    index = str.indexOf(searchStr, startIndex);
  }

  return indices;
};

const getEntityRanges = (
  text: string,
  mentionName: string,
  mentionKey: number
) => {
  const indices = getIndicesOf(mentionName, text);

  if (indices.length > 0) {
    return indices.map(offset => ({
      key: mentionKey,
      length: mentionName.length,
      offset,
    }));
  }

  return null;
};

const convertStringToEditorState = (text: string, tags: string[]) => {
  const rawContent = convertToRaw(ContentState.createFromText(text));

  const rawState = tags.map(tag => ({
    type: 'mention',
    mutability: 'IMMUTABLE',
    data: tag,
  }));

  (rawContent as any).entityMap = [...rawState];

  rawContent.blocks = rawContent.blocks.map(block => {
    const ranges: {
      key: number;
      length: number;
      offset: number;
    }[] = [];

    tags.forEach((tag, index) => {
      const entityRanges = getEntityRanges(block.text, `@${tag}`, index);
      if (entityRanges) {
        ranges.push(...entityRanges);
      }
    });

    return { ...block, entityRanges: ranges };
  });

  const contentState = convertFromRaw(rawContent);

  return EditorState.createWithContent(contentState);
};

export default convertStringToEditorState;
