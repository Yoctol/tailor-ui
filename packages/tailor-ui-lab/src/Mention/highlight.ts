export const applyHighlights = ({
  suggestions,
  value,
  highlightInvalid,
}: {
  suggestions: string[];
  value: string;
  highlightInvalid: boolean;
}) => {
  const transformedValue = value.replace(/\n$/g, '\n\n');
  const mentionRegexp = /\{{2}[^{}\n]+\}{2}/g;
  const allSuggestions = suggestions.join('|');

  const replacer = (match: string) => {
    let valid = true;

    if (highlightInvalid) {
      const validRegexp = new RegExp(`{{2}(${allSuggestions})}{2}`, 'g');
      valid = validRegexp.test(match);
    }

    return `<span class="mention-highlight ${
      !valid ? 'invalid' : ''
    }">${match}</span>`;
  };

  return transformedValue.replace(mentionRegexp, replacer);
};
