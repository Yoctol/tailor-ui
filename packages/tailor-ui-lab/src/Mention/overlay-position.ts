interface GetOverlayPosition {
  mentionRect: {
    top: number;
    left: number;
  };
  coordinates: {
    top: number;
    left: number;
    height: number;
  };
  target: HTMLElement;
  placement: 'bottom' | 'top';
}

export interface OverlayPosition {
  top: number;
  left: number;
}

export const getOverlayPosition = ({
  mentionRect,
  coordinates,
  target,
  placement,
}: GetOverlayPosition): OverlayPosition => {
  const top =
    mentionRect.top +
    coordinates.top -
    target.scrollTop +
    (placement === 'bottom' ? coordinates.height : 0);
  const left = mentionRect.left + coordinates.left - target.scrollLeft;

  return {
    top,
    left,
  };
};
