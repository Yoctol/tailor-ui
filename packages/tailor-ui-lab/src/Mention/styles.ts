import styled, { css } from 'styled-components';
import { animated } from 'react-spring';
import { rgba } from 'polished';

import { Textarea as BaseTextarea, inputStyles } from 'tailor-ui';

export const Textarea = styled(BaseTextarea)`
  position: absolute;
  top: 0;
  background-color: transparent;
`;

export const MentionWrapper = styled.div<{ disabled?: boolean }>`
  position: relative;

  .mention-highlight {
    display: inline-flex;
    background-color: ${p =>
      p.disabled ? p.theme.colors.gray400 : rgba(p.theme.colors.success, 0.3)};
    line-height: 1.3;
    text-decoration: none;

    &.start {
      border-top-left-radius: ${p => p.theme.radii.base};
      border-bottom-left-radius: ${p => p.theme.radii.base};
    }

    &.end {
      border-top-right-radius: ${p => p.theme.radii.base};
      border-bottom-right-radius: ${p => p.theme.radii.base};
    }

    &.invalid {
      background-color: ${p =>
        p.disabled ? p.theme.colors.gray400 : rgba(p.theme.colors.danger, 0.3)};
    }
  }
`;

export const Highlights = styled.div`
  ${inputStyles};

  /* stylelint-disable-next-line order/properties-order */
  border-color: transparent;
  color: transparent;
  word-break: break-word;
  white-space: pre-wrap;
`;

export const SuggestionList = styled(animated.div)`
  display: flex;
  z-index: 2;
  box-sizing: border-box;
  flex-direction: column;
  min-width: 120px;
  max-height: 200px;
  margin-top: ${p => p.theme.space[1]};
  padding: ${p => p.theme.space[2]} 0;
  overflow-y: auto;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray300};
  outline: none;
  background-color: ${p => p.theme.colors.light};
  box-shadow: ${p => p.theme.shadows.lg};
  cursor: pointer;
`;

export const SuggestionItem = styled.div<{ active: boolean }>`
  display: inline-flex;
  flex: none;
  align-items: center;
  height: ${p => p.theme.heights.base};
  padding: 0 ${p => p.theme.paddings.sm};
  font-size: ${p => p.theme.fontSizes.base};
  text-overflow: ellipsis;
  white-space: nowrap;
  transition: background-color 0.4s cubic-bezier(0.27, 1.27, 0.48, 0.56);

  &:hover {
    background-color: #cce7ff;
  }

  ${({ active }) =>
    active &&
    css`
      background-color: #e6f3ff;
    `}
`;
