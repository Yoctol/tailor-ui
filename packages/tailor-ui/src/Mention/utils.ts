interface GetOverlayPosition {
  mentionRect: {
    top: number;
    left: number;
  };
  coordinates: {
    top: number;
    left: number;
    height: number;
  };
  target: HTMLElement;
  placement: 'bottom' | 'top';
}

export interface OverlayPosition {
  top: number;
  left: number;
}

export const getOverlayPosition = ({
  mentionRect,
  coordinates,
  target,
  placement,
}: GetOverlayPosition): OverlayPosition => {
  const top =
    mentionRect.top +
    coordinates.top -
    target.scrollTop +
    (placement === 'bottom' ? coordinates.height : 0);
  const left = mentionRect.left + coordinates.left - target.scrollLeft;

  return {
    top,
    left,
  };
};

const PREFIX = '{{';
const SUFFIX = '}}';

export const getCursorPosition = (value: string, selectionStart: number) => {
  const startPos =
    value.lastIndexOf(PREFIX, selectionStart) > -1
      ? value.lastIndexOf(PREFIX, selectionStart) + 2
      : -1;

  const suffixPos = value.indexOf(SUFFIX, selectionStart);
  const nextPrefixPos = value.indexOf(PREFIX, selectionStart);
  const spacePos = value.indexOf(' ', selectionStart);

  const positions = [suffixPos, nextPrefixPos, spacePos]
    .filter((pos) => pos !== -1)
    .sort((a, b) => a - b);

  if (
    selectionStart === startPos &&
    nextPrefixPos !== -1 &&
    nextPrefixPos === positions[0]
  ) {
    return {
      startPos,
      endPos: startPos,
    };
  }

  const endPos = positions[0] || selectionStart;

  return {
    startPos,
    endPos,
  };
};

export const getMentionCursor = (
  originValue: string,
  selectionStart: number
) => {
  const value = originValue.replace(/[\r\n]/g, ' ');

  const { startPos, endPos } = getCursorPosition(value, selectionStart);
  const mention = value.substring(startPos, endPos);
  const searchValue = value.substring(startPos, selectionStart);
  const createValue =
    mention.length > searchValue.length ? mention : searchValue;

  if (
    startPos < 0 ||
    endPos < 0 ||
    mention.includes('{') ||
    mention.includes('}') ||
    selectionStart < startPos ||
    selectionStart > endPos
  ) {
    return {
      mention: null,
      searchValue: null,
      createValue: null,
      startPos: -1,
      endPos: -1,
    };
  }

  return {
    mention,
    searchValue,
    createValue,
    startPos,
    endPos,
  };
};
