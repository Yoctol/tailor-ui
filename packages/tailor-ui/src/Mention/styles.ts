import styled from 'styled-components';
import { animated } from 'react-spring';

import { Textarea as BaseTextarea } from '../Input/Textarea';
import { inputStyles } from '../Input/styles';

export const Textarea = styled(BaseTextarea)`
  position: absolute;
  top: 0;
  background-color: transparent;
  color: transparent;
  caret-color: ${p => p.theme.colors.gray700};
`;

export const MentionWrapper = styled.div<{ disabled?: boolean }>`
  position: relative;

  .mention-highlight {
    display: inline-flex;
    color: ${p =>
      p.disabled ? p.theme.colors.gray400 : p.theme.colors.primaryLight};
    line-height: 1.3;
    text-decoration: none;

    &.invalid {
      color: ${p =>
        p.disabled ? p.theme.colors.gray400 : p.theme.colors.danger};
    }
  }
`;

export const Highlights = styled.div`
  ${inputStyles};

  /* stylelint-disable-next-line order/properties-order */
  overflow-y: auto;
  border-color: transparent;
  color: ${p => p.theme.colors.gray700};
  word-break: break-word;
  white-space: pre-wrap;
  transition: none;
`;

export const SuggestionList = styled(animated.div)`
  display: flex;
  z-index: 2;
  box-sizing: border-box;
  flex-direction: column;
  min-width: 120px;
  max-height: 180px;
  overflow-y: auto;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.lg};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.light};
  box-shadow: ${p => p.theme.shadows.base};
  color: ${p => p.theme.colors.gray700};
  font-size: ${p => p.theme.fontSizes.sm};
  text-align: left;
  white-space: nowrap;
  cursor: pointer;
`;

export const SuggestionItem = styled.div<{ active: boolean }>`
  display: flex;
  flex: none;
  align-items: center;
  height: 36px;
  padding: 0 24px;
  background-color: ${p =>
    p.active ? p.theme.colors.primaryLight : p.theme.colors.light};
  color: ${p => (p.active ? p.theme.colors.light : p.theme.colors.gray700)};
  cursor: pointer;

  &:hover {
    background-color: ${p => p.theme.colors.primaryLight};
    color: ${p => p.theme.colors.light};
  }
`;
