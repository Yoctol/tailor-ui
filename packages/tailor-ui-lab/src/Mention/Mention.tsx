import React, {
  ChangeEvent,
  FunctionComponent,
  KeyboardEvent,
  Reducer,
  UIEvent,
  useEffect,
  useReducer,
  useRef,
} from 'react';

import Suggestions from './Suggestions';
import { Highlights, MentionWrapper, Textarea } from './styles';
import { OverlayPosition, getMentionCursor, getOverlayPosition } from './utils';
import { applyHighlights } from './highlight';
import { getCaretCoordinates } from './textarea-caret-position';

const placement = 'bottom';

const LEFT_ARROW = 37;
const RIGHT_ARROW = 39;
const UP_ARROW = 38;
const DOWN_ARROW = 40;
const ENTER = 13;
const ESCAPE = 27;
const TAB = 9;

type FilteredSuggestionType = 'option' | 'create';

export interface FilteredSuggestion {
  type: FilteredSuggestionType;
  value: string;
}

export type FilteredSuggestions = FilteredSuggestion[];

interface MentionReducerState {
  value: string;
  height: number;
  activeIndex: number;
  caretPos: number;
  dropdownVisible: boolean;
  filteredSuggestions: FilteredSuggestions;
  overlayPosition: OverlayPosition;
}

interface CursorMention {
  mention: string | null;
  searchValue: string | null;
  startPos: number;
  endPos: number;
}

interface MentionReducerAction {
  type:
    | 'closeDropdown'
    | 'openDropdown'
    | 'selectSuggestion'
    | 'setPreviousItemActive'
    | 'setNextItemActive'
    | 'setSpecfiedItemActive'
    | 'updateValue'
    | 'updateHeight'
    | 'updateCaretPos';
  payload?: any;
}

interface InitialState {
  value: string;
  suggestions: string[];
}

const initialState = ({ value, suggestions }: InitialState) => ({
  value,
  height: 0,
  activeIndex: -1,
  caretPos: -1,
  dropdownVisible: false,
  filteredSuggestions: suggestions.map(suggestion => ({
    type: 'option' as FilteredSuggestionType,
    value: suggestion,
  })),
  overlayPosition: {
    top: 0,
    left: 0,
  },
});

const reducer: Reducer<MentionReducerState, MentionReducerAction> = (
  state,
  action
) => {
  switch (action.type) {
    case 'closeDropdown':
      return {
        ...state,
        activeIndex: -1,
        dropdownVisible: false,
      };
    case 'openDropdown':
      return {
        ...state,
        activeIndex: 0,
        dropdownVisible: true,
        ...action.payload,
      };
    case 'selectSuggestion':
      return {
        ...state,
        dropdownVisible: false,
        activeIndex: -1,
        ...action.payload,
      };
    case 'setPreviousItemActive':
      return {
        ...state,
        activeIndex:
          state.activeIndex - 1 < 0
            ? state.filteredSuggestions.length - 1
            : state.activeIndex - 1,
      };
    case 'setNextItemActive':
      return {
        ...state,
        activeIndex:
          state.activeIndex + 1 <= state.filteredSuggestions.length - 1
            ? state.activeIndex + 1
            : 0,
      };
    case 'updateValue':
      return {
        ...state,
        value: action.payload,
      };
    case 'updateHeight':
      return {
        ...state,
        height: action.payload,
      };
    case 'updateCaretPos':
      return {
        ...state,
        caretPos: action.payload,
      };
    default:
      throw new Error();
  }
};

type ResetDropdownType =
  | KeyboardEvent<HTMLTextAreaElement>
  | ChangeEvent<HTMLTextAreaElement>;

export type FormatCreateText = (text: string) => string;

interface MentionProps {
  value?: string;
  defaultValue: string;
  suggestions: string[];
  disabled?: boolean;
  creatable: boolean;
  highlightInvalid: boolean;
  formatCreateText: FormatCreateText;
  onBlur?: (value: string) => void;
  onChange?: (value: string) => void;
  onMentionCreate: (newMention: string) => void;
}

const Mention: FunctionComponent<MentionProps> = ({
  value: valueFromProps,
  defaultValue,
  suggestions,
  onBlur,
  onChange,
  onMentionCreate,
  disabled,
  creatable,
  highlightInvalid,
  formatCreateText,
  ...props
}) => {
  const mentionRef = useRef<any>(null);
  const highlightRef = useRef<any>(null);

  const cursorMention = useRef<CursorMention>({
    mention: null,
    searchValue: null,
    startPos: -1,
    endPos: -1,
  });

  const [state, dispatch] = useReducer(
    reducer,
    initialState({
      value: valueFromProps || defaultValue || '',
      suggestions,
    })
  );

  const value = valueFromProps || state.value;

  useEffect(() => {
    dispatch({
      type: 'updateHeight',
      payload: mentionRef.current.textarea.offsetHeight,
    });
  }, []);

  useEffect(() => {
    if (state.caretPos !== -1 && mentionRef.current.textarea) {
      mentionRef.current.textarea.focus();
      mentionRef.current.textarea.setSelectionRange(
        state.caretPos,
        state.caretPos
      );

      dispatch({ type: 'updateCaretPos', payload: -1 });
    }
  }, [state.caretPos]);

  const resetDropdown = ({ currentTarget }: ResetDropdownType) => {
    const cursor = getMentionCursor(
      currentTarget.value,
      currentTarget.selectionStart
    );

    cursorMention.current = cursor;

    if (cursor.mention !== null) {
      const filteredSuggestions = [
        ...suggestions
          .filter(suggestion => suggestion.includes(cursor.searchValue))
          .map(suggestion => ({
            type: 'option',
            value: suggestion,
          })),
        ...(creatable &&
        cursor.createValue !== '' &&
        !suggestions.includes(cursor.createValue)
          ? [{ type: 'create', value: cursor.createValue }]
          : []),
      ];

      if (filteredSuggestions.length === 0) {
        if (state.dropdownVisible) {
          dispatch({ type: 'closeDropdown' });
        }

        return;
      }

      const coordinates = getCaretCoordinates(
        currentTarget,
        cursor.startPos - 2
      );

      const mentionRect = currentTarget.getBoundingClientRect();

      const overlayPosition = getOverlayPosition({
        coordinates,
        mentionRect,
        target: currentTarget,
        placement,
      });

      dispatch({
        type: 'openDropdown',
        payload: {
          overlayPosition,
          filteredSuggestions,
        },
      });
    } else if (state.dropdownVisible) {
      dispatch({ type: 'closeDropdown' });
    }
  };

  const selectSuggestion = (suggestion: FilteredSuggestion) => {
    const { startPos, endPos } = cursorMention.current;
    const insertValue = `{{${suggestion.value}}}`;
    const newValue = [
      value.slice(0, startPos - 2),
      insertValue,
      value.slice(endPos + (value[endPos] === '}' ? 2 : 0), value.length),
    ].join('');

    dispatch({
      type: 'selectSuggestion',
      payload: {
        caretPos: startPos - 2 + insertValue.length,
        value: newValue,
      },
    });

    if (suggestion.type === 'create') {
      onMentionCreate(suggestion.value);
    }

    if (onChange) {
      onChange(newValue);
    }
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLTextAreaElement>) => {
    const { keyCode } = event;

    if (keyCode === LEFT_ARROW || keyCode === RIGHT_ARROW) {
      event.stopPropagation();
      resetDropdown(event);
    }
  };

  const handleKeyDown = (event: KeyboardEvent<HTMLTextAreaElement>) => {
    const { dropdownVisible, activeIndex, filteredSuggestions } = state;
    const { keyCode } = event;

    if (
      dropdownVisible &&
      keyCode === ENTER &&
      activeIndex !== -1 &&
      filteredSuggestions.length &&
      cursorMention.current
    ) {
      event.preventDefault();
      event.stopPropagation();

      selectSuggestion(filteredSuggestions[activeIndex]);

      return;
    }

    if (dropdownVisible && [TAB, ESCAPE].includes(keyCode)) {
      dispatch({ type: 'closeDropdown' });

      return;
    }

    if (dropdownVisible && keyCode === UP_ARROW) {
      event.preventDefault();
      event.stopPropagation();

      dispatch({ type: 'setPreviousItemActive' });

      return;
    }

    if (dropdownVisible && keyCode === DOWN_ARROW) {
      event.preventDefault();
      event.stopPropagation();

      dispatch({ type: 'setNextItemActive' });
    }
  };

  const handleOnBlur = (event: ChangeEvent<HTMLTextAreaElement>) => {
    const { value: newValue } = event.currentTarget;

    dispatch({ type: 'closeDropdown' });

    if (onBlur) {
      onBlur(newValue);
    }
  };

  const handleChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
    const { value: newValue } = event.currentTarget;

    if (newValue !== value) {
      dispatch({ type: 'updateValue', payload: newValue });

      if (onChange) {
        onChange(newValue);
      }
    }

    resetDropdown(event);
  };

  return (
    <MentionWrapper disabled={disabled}>
      <Highlights
        ref={highlightRef}
        style={{ height: state.height }}
        dangerouslySetInnerHTML={{
          __html: applyHighlights({ suggestions, value, highlightInvalid }),
        }}
      />
      <Textarea
        ref={mentionRef}
        onKeyUp={handleKeyUp}
        onKeyDown={handleKeyDown}
        value={value}
        onClick={resetDropdown}
        onBlur={handleOnBlur}
        onChange={handleChange}
        disabled={disabled}
        onScroll={(event: UIEvent<HTMLTextAreaElement>) => {
          if (highlightRef.current) {
            highlightRef.current.scrollTop = event.currentTarget.scrollTop;
          }
          if (state.dropdownVisible) {
            dispatch({ type: 'closeDropdown' });
          }
        }}
        onResize={(event: ChangeEvent<HTMLTextAreaElement>) => {
          // https://github.com/buildo/react-autosize-textarea/issues/109#issuecomment-430002058
          setTimeout(() =>
            dispatch({
              type: 'updateHeight',
              payload: event.target.offsetHeight,
            })
          );
        }}
        {...props}
      />
      <Suggestions
        formatCreateText={formatCreateText}
        dropdownVisible={state.dropdownVisible}
        overlayPosition={state.overlayPosition}
        filteredSuggestions={state.filteredSuggestions}
        activeIndex={state.activeIndex}
        onSuggestionClick={selectSuggestion}
      />
    </MentionWrapper>
  );
};

Mention.defaultProps = {
  defaultValue: '',
  suggestions: [],
  creatable: false,
  highlightInvalid: false,
  onMentionCreate: () => {},
  formatCreateText: createValue =>
    `Press Enter to create mention: ${createValue}`,
};

export default Mention;
