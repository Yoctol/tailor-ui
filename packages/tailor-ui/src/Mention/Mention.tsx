import Mention, {
  Nav,
  getMentions,
  toEditorState,
  toString,
} from 'rc-editor-mention';
import React, {
  FocusEvent,
  FocusEventHandler,
  PureComponent,
  RefObject,
  createRef,
} from 'react';
import styled, { css } from 'styled-components';

import tag from '../utils/CleanTag';
import { Size, inputStyles } from '../Input';

export interface MentionWrapperProps {
  focus?: boolean;
  disabled?: boolean;
  error?: boolean;
  size: Size;
}

const MentionWrapper = styled(tag.div)<MentionWrapperProps>`
  position: relative;

  .yoctol-ui-mention-wrapper {
    display: inline-block;
    position: relative;
    width: 100%;
    vertical-align: middle;

    .yoctol-ui-mention-editor {
      ${inputStyles};

      /* stylelint-disable-next-line order/properties-order */
      height: auto;

      ${p =>
        p.focus &&
        !p.disabled &&
        css`
          border-color: ${p.theme.colors.gray700};
        `};

      ${p =>
        p.error &&
        css`
          border-color: ${p.theme.colors.danger};
        `};

      ${p =>
        p.disabled &&
        css`
          border-color: ${p.theme.colors.gray500};
          opacity: 0.5;
          background-color: ${p.theme.colors.gray300};
          cursor: not-allowed;
        `};

      div > span:not([data-offset-key]) {
        color: ${p => p.theme.colors.success};
      }
    }

    &.multilines .yoctol-ui-mention-editor {
      height: auto;
    }
  }

  .yoctol-ui-mention {
    &-dropdown {
      display: flex;
      position: absolute;
      z-index: 2;
      top: -9999px;
      left: -9999px;
      flex-direction: column;
      min-width: 120px;
      max-height: 200px;
      margin-top: 1.5em;
      overflow-x: hidden;
      overflow-y: auto;
      border: ${p => p.theme.borders.base};
      border-radius: ${p => p.theme.radii.sm};
      border-color: ${p => p.theme.colors.gray300};
      outline: none;
      background-color: ${p => p.theme.colors.light};
      box-shadow: ${p => p.theme.shadows.lg};
      line-height: 1;
      cursor: pointer;

      &-item {
        display: flex;
        position: relative;
        flex: none;
        align-items: center;
        height: ${p => p.theme.heights.base};
        padding: 0 ${p => p.theme.paddings.sm};
        overflow: hidden;
        font-weight: normal;
        line-height: 22px;
        text-overflow: ellipsis;
        white-space: nowrap;
        cursor: pointer;
        ${p => p.theme.transition};

        &:hover {
          background-color: ${p => p.theme.colors.gray300};
        }

        &.focus,
        &-active {
          background-color: ${p => p.theme.colors.gray200};
        }

        &-disabled {
          opacity: 0.5;
          cursor: not-allowed;
        }

        &-selected {
          &,
          &:hover {
            background-color: ${p => p.theme.colors.gray300};
          }
        }

        &-divider {
          height: 1px;
          margin: 1px 0;
          overflow: hidden;
          background-color: ${p => p.theme.colors.gray100};
          line-height: 0;
        }
      }
    }
  }
`;

const MentionTag = styled.span.attrs({
  contentEditable: false,
})`
  padding-left: ${p => p.theme.space[1]};
  color: ${p => p.theme.colors.success};
`;

export interface MentionEditorProps {
  autoFocus?: boolean;
  /**
   * The text
   */
  createText?: string;
  /**
   * Tell if the input is disabled.
   */
  disabled?: boolean;
  /**
   * Set the mention to error status
   */
  error?: boolean;
  /**
   * placeholder of input
   */
  placeholder?: string;
  /**
   * character which will trigger Mention to show mention list
   */
  prefix?: string;
  /**
   * The size of the input
   */
  size: 'sm' | 'md' | 'lg';
  /**
   * suggestion content
   */
  suggestions: string[];
  /**
   * textarea mode
   */
  textarea?: boolean;
  /**
   * Callback function called when mention component blur
   */
  onBlur?: FocusEventHandler<HTMLElement>;
  /**
   * Callback function called when content of input changes
   */
  onChange?: (contentState: any) => any;
  /**
   * Callback function called when mention component get focus
   */
  onFocus?: FocusEventHandler<HTMLElement>;
  /**
   * Callback function called when select from suggestions
   */
  onSelect?: (suggestion: string, data?: any) => any;
}

class MentionEditor extends PureComponent<MentionEditorProps> {
  static toString = toString;

  static toContentState = toEditorState;

  static getMentions = getMentions;

  static defaultProps = {
    autoFocus: false,
    createText: 'Press enter to create new mention: ',
    disabled: false,
    error: false,
    suggestions: [],
    placeholder: null,
    prefix: '@',
    size: 'md',
    textarea: false,
  };

  mentionRef: RefObject<any> = createRef();

  mentionWrapperRef?: RefObject<any>;

  state = {
    suggestions: this.props.suggestions,
    focus: false,
  };

  componentDidMount() {
    if (this.props.autoFocus) {
      this.focus();
    }
  }

  onSearchChange = (value: string) => {
    const searchValue = value.toLowerCase();
    const filtered = this.props.suggestions.filter(
      suggestion => suggestion.toLowerCase().indexOf(searchValue) !== -1
    );

    const createTag = [
      <Nav value={value}>
        {this.props.createText}
        <MentionTag>{value}</MentionTag>
      </Nav>,
    ];

    const suggestions =
      filtered.length !== 0
        ? [...filtered, ...(value !== '' ? createTag : [])]
        : createTag;

    this.setState({
      suggestions,
    });
  };

  onFocus = (event: FocusEvent<HTMLElement>) => {
    this.setState(() => ({
      focus: true,
    }));

    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
  };

  onBlur = (event: FocusEvent<HTMLElement>) => {
    this.setState(() => ({
      focus: false,
    }));

    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  };

  focus = () => {
    this.mentionRef.current._editor.focusEditor();
  };

  render() {
    const { suggestions, focus } = this.state;
    const { size, error, disabled, textarea } = this.props;

    return (
      <MentionWrapper
        size={size}
        error={error}
        disabled={disabled}
        focus={focus}
        ref={(ref: any) => {
          this.mentionWrapperRef = ref;
        }}
      >
        <Mention
          {...this.props}
          prefixCls="yoctol-ui-mention"
          mode="mutable"
          ref={this.mentionRef}
          getSuggestionContainer={() => this.mentionWrapperRef}
          tag={MentionTag}
          prefix={this.props.prefix}
          suggestions={suggestions}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onSearchChange={this.onSearchChange}
          multiLines={textarea}
        />
      </MentionWrapper>
    );
  }
}

export default MentionEditor;
