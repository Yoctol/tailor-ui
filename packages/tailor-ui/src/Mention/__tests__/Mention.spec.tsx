import React from 'react';

import { fireEvent, render } from 'test/test-utils';

import { Mention } from '../Mention';

it('should be defined', () => {
  expect(Mention).toBeDefined();
});

describe('Mention', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Mention
        defaultValue=""
        suggestions={[]}
        creatable
        highlightInvalid={false}
        formatCreateText={() => 'Press Enter'}
        onMentionCreate={() => {}}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should trigger onBlur from props when suggestions is visible and click outside mention', async () => {
    const onBlurFromProps = jest.fn();
    const onChangeFromProps = jest.fn();

    const MENTION_TEST_ID = 'my-mention';
    const OTHER_TEST_ID = 'your-dom';

    const { getByTestId } = render(
      <div>
        <Mention
          data-testid={MENTION_TEST_ID}
          defaultValue=""
          suggestions={['drink']}
          creatable
          highlightInvalid={false}
          formatCreateText={() => 'Press Enter'}
          onMentionCreate={() => {}}
          onBlur={onBlurFromProps}
          onChange={onChangeFromProps}
        />
        <div data-testid={OTHER_TEST_ID} />
      </div>
    );

    const mention = getByTestId(MENTION_TEST_ID);
    const other = getByTestId(OTHER_TEST_ID);

    fireEvent.focus(mention);
    fireEvent.change(mention, {
      target: {
        value: '{{',
        selectionStart: 2,
      },
    });

    fireEvent.click(other);

    expect(onChangeFromProps).toBeCalledWith('{{');
    expect(onBlurFromProps).toBeCalled();
  });
});
