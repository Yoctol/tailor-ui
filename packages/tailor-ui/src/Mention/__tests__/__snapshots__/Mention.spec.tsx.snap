// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mention should render correctly 1`] = `
.c2 {
  word-wrap: break-word;
  resize: none;
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
  height: auto;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
}

.c2:hover {
  border-color: #3c5ad0;
}

.c2:focus {
  border-color: #3c5ad0;
  box-shadow: inset 0 0 0 2px #e8ecf5;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #bfbfbf;
  background-color: #e0e0e0;
  color: #8b8b8b;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c2::-moz-placeholder {
  color: #bfbfbf;
}

.c2:-ms-input-placeholder {
  color: #bfbfbf;
}

.c2::placeholder {
  color: #bfbfbf;
}

.c3 {
  position: absolute;
  top: 0;
  background-color: transparent;
  color: transparent;
  caret-color: #373737;
}

.c0 {
  position: relative;
}

.c0 .mention-highlight {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #6B88c5;
  line-height: 1.3;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 .mention-highlight.invalid {
  color: #fc5475;
}

.c1 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
  overflow-y: auto;
  border-color: transparent;
  color: #373737;
  word-break: break-word;
  white-space: pre-wrap;
  -webkit-transition: none;
  transition: none;
}

.c1:hover {
  border-color: #3c5ad0;
}

.c1:focus {
  border-color: #3c5ad0;
  box-shadow: inset 0 0 0 2px #e8ecf5;
}

.c1:disabled,
.c1 [disabled] {
  border-color: #bfbfbf;
  background-color: #e0e0e0;
  color: #8b8b8b;
  cursor: not-allowed;
}

.c1::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c1::-moz-placeholder {
  color: #bfbfbf;
}

.c1:-ms-input-placeholder {
  color: #bfbfbf;
}

.c1::placeholder {
  color: #bfbfbf;
}

<div
  class="c0"
>
  <div
    class="c1"
    style="height: 0px;"
  />
  <textarea
    class="c2 c3"
    rows="1"
    style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
  />
</div>
`;
